primitives/src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Merkle Tree traits and implementations
+pub mod append_only;
+pub mod examples;
+pub mod macros;
+pub mod sparse_merkle_tree;
+
+mod internal;
+
+use crate::errors::PrimitivesError;
+use ark_std::{borrow::Borrow, fmt::Debug, string::ToString, vec, vec::Vec};
+use num_bigint::BigUint;
+use num_traits::ToPrimitive;
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+/// The result of querying at an index in the tree
+/// Typically, F for element type, P for membership proof type and N for
+/// non-membership proof type
+pub enum LookupResult<F, P, N> {
+    /// The value at the given index, and a proof of validity
+    Ok(F, P),
+    /// The index is valid but we do not have the leaf in memory
+    NotInMemory,
+    /// The index is outside the occupied range in the tree, and a
+    /// non-membership proof
+    EmptyLeaf(N),
```

Comment:
I feel like `Leaf` is a bit of an implementation detail. If this is going to be in a public, generic interface, maybe something higher-level like `NotInRange` or `NotFound` would be better?